### ISSUE TITLE
Many stacktraces when running on GNOME 3.26 in extension.js:256

### ISSUE THREAD

**Issue by genodeftest** [sustmi/gnome-shell-extension-windowoverlay-icons#4]
Steps to reproduce:
unclear. Have gnome-shell with this extension run for a while.

What happens:
This stacktrace is written to syslog:
```
== Stack trace for context 0x55c147d90000 ==
#0 0x7fff8b6c3e90 b   /home/username/.local/share/gnome-shell/extensions/windowoverlay-icons@sustmidown.centrum.cz/extension.js:256 (0x7fd9a6890808 @ 123)
#1 0x55c148222450 i   resource:///org/gnome/shell/ui/extensionSystem.js:82 (0x7fd9c58b4230 @ 431)
#2 0x55c1482223d0 i   resource:///org/gnome/shell/ui/extensionSystem.js:344 (0x7fd9c58b4cd0 @ 13)
#3 0x7fff8b6c4cb0 b   self-hosted:251 (0x7fd9cb1d25e8 @ 223)
#4 0x55c148222350 i   resource:///org/gnome/shell/ui/extensionSystem.js:343 (0x7fd9c58b4c48 @ 63)
#5 0x55c1482222d0 i   resource:///org/gnome/shell/ui/extensionSystem.js:361 (0x7fd9c58b4d58 @ 87)
#6 0x7fff8b6c61d0 b   resource:///org/gnome/gjs/modules/signals.js:126 (0x7fd9cb1e2b38 @ 386)
#7 0x7fff8b6c69e0 b   resource:///org/gnome/shell/ui/sessionMode.js:204 (0x7fd9c4ca3098 @ 254)
#8 0x7fff8b6c76e0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7fd9cb1c2bc0 @ 71)
#9 0x55c148222190 i   resource:///org/gnome/shell/ui/sessionMode.js:166 (0x7fd9c4c9ede0 @ 40)
#10 0x7fff8b6c82e0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7fd9cb1c2bc0 @ 71)
#11 0x55c1482220e8 i   resource:///org/gnome/shell/ui/screenShield.js:1281 (0x7fd9c4c76098 @ 188)
#12 0x7fff8b6c8ee0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7fd9cb1c2bc0 @ 71)
#13 0x55c148222060 i   resource:///org/gnome/shell/ui/screenShield.js:899 (0x7fd9c4c741a8 @ 18)
#14 0x7fff8b6c9ae0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7fd9cb1c2bc0 @ 71)
#15 0x7fff8b6c9ba0 b   self-hosted:915 (0x7fd9cb1ee5e8 @ 367)
#16 0x7fff8b6c9c90 b   resource:///org/gnome/gjs/modules/signals.js:126 (0x7fd9cb1e2b38 @ 386)
#17 0x55c148221fd8 i   resource:///org/gnome/shell/ui/lightbox.js:186 (0x7fd9c5886c48 @ 33)
#18 0x7fff8b6ca9e0 b   self-hosted:913 (0x7fd9cb1ee5e8 @ 346)
#19 0x55c148221f50 i   resource:///org/gnome/shell/ui/tweener.js:91 (0x7fd9cb1df098 @ 37)
#20 0x55c148221eb0 i   resource:///org/gnome/gjs/modules/tweener/tweener.js:203 (0x7fd9cb1e9cd0 @ 54)
#21 0x55c148221d90 i   resource:///org/gnome/gjs/modules/tweener/tweener.js:332 (0x7fd9cb1e9d58 @ 1626)
#22 0x55c148221d00 i   resource:///org/gnome/gjs/modules/tweener/tweener.js:345 (0x7fd9cb1e9de0 @ 100)
#23 0x55c148221c88 i   resource:///org/gnome/gjs/modules/tweener/tweener.js:360 (0x7fd9cb1e9e68 @ 10)
#24 0x7fff8b6cd490 b   resource:///org/gnome/gjs/modules/signals.js:126 (0x7fd9cb1e2b38 @ 386)
#25 0x55c148221bf8 i   resource:///org/gnome/shell/ui/tweener.js:208 (0x7fd9cb1df808 @ 159)
#26 0x7fff8b6ce100 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7fd9cb1c2bc0 @ 71)
#27 0x55c148221b70 i   resource:///org/gnome/shell/ui/tweener.js:183 (0x7fd9cb1df780 @ 20)
#28 0x7fff8b6cee00 b   self-hosted:917 (0x7fd9cb1ee5e8 @ 394)
```
or this other stacktrace:
```
#0 0x7ffc203c23d0 b   /home/username/.local/share/gnome-shell/extensions/windowoverlay-icons@sustmidown.centrum.cz/extension.js:256 (0x7f5a85d90808 @ 123)
#1 0x55d4afb3ef70 i   resource:///org/gnome/shell/ui/extensionSystem.js:70 (0x7f5aa4db4230 @ 229)
#2 0x55d4afb3eee8 i   resource:///org/gnome/shell/ui/extensionSystem.js:184 (0x7f5aa4db4450 @ 18)
#3 0x55d4afb3ee30 i   resource:///org/gnome/shell/ui/extensionSystem.js:199 (0x7f5aa4db44d8 @ 81)
#4 0x55d4afb3ed90 i   resource:///org/gnome/shell/ui/extensionSystem.js:299 (0x7f5aa4db4918 @ 76)
```
or this one:
```
== Stack trace for context 0x557564bde000 ==
#0 0x7ffc1cc13d60 b   /home/username/.local/share/gnome-shell/extensions/windowoverlay-icons@sustmidown.centrum.cz/extension.js:256 (0x7f001bf90808 @ 123)
#1 0x55756501d098 i   resource:///org/gnome/shell/ui/extensionSystem.js:82 (0x7f003afb4230 @ 431)
#2 0x55756501d018 i   resource:///org/gnome/shell/ui/extensionSystem.js:344 (0x7f003afb4cd0 @ 13)
#3 0x7ffc1cc14b80 b   self-hosted:251 (0x7f00489d25e8 @ 223)
#4 0x55756501cf98 i   resource:///org/gnome/shell/ui/extensionSystem.js:343 (0x7f003afb4c48 @ 63)
#5 0x55756501cf18 i   resource:///org/gnome/shell/ui/extensionSystem.js:361 (0x7f003afb4d58 @ 87)
#6 0x7ffc1cc160a0 b   resource:///org/gnome/gjs/modules/signals.js:126 (0x7f00489e2b38 @ 386)
#7 0x7ffc1cc168b0 b   resource:///org/gnome/shell/ui/sessionMode.js:204 (0x7f003a3a4098 @ 254)
#8 0x7ffc1cc175b0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7f00489c2bc0 @ 71)
#9 0x55756501cdd8 i   resource:///org/gnome/shell/ui/sessionMode.js:166 (0x7f003a39ede0 @ 40)
#10 0x7ffc1cc181b0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7f00489c2bc0 @ 71)
#11 0x55756501cd30 i   resource:///org/gnome/shell/ui/screenShield.js:1281 (0x7f003a376098 @ 188)
#12 0x7ffc1cc18db0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7f00489c2bc0 @ 71)
#13 0x55756501cc80 i   resource:///org/gnome/shell/ui/screenShield.js:1327 (0x7f003a376120 @ 391)
#14 0x7ffc1cc199b0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7f00489c2bc0 @ 71)
#15 0x55756501cbf0 i   resource:///org/gnome/shell/misc/systemActions.js:403 (0x7f003a4b63c8 @ 80)
#16 0x7ffc1cc1a5b0 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7f00489c2bc0 @ 71)
#17 0x55756501cb70 i   resource:///org/gnome/shell/ui/status/system.js:287 (0x7f003a4ab5e8 @ 53)
```
Each of these stacktraces have in common that they happen several tens or hundreds of times per second and start with the same first line in `extension.js:256`. This is the `i.destroy();` line in the `disable()` function.

What should happen:
No stacktrace.

Additional info:
Printing these stacktraces is a feature introduced in or after 3.26.0 release, afaik.

Please note that I did not disable this extension at all and I don't know why this function gets called. I guess that it is being called on screen lock or suspend.

Installed software versions:
gnome-shell-3.26.2-4.fc27.x86_64
gjs-1.50.4-1.fc27.x86_64
mutter-3.26.2-2.fc27.x86_64

**Comment by genodeftest** [comment]
There are similar issues in gnome-shell's code itself, see e.g. https://gitlab.gnome.org/GNOME/gnome-shell/issues/11

It is unclear, why those stacktraces are thrown. They might be indicating that the object has already been destroyed somewhere else or that it cannot be destroyed because it is still being referenced by from somewhere else.

**Comment by genodeftest** [comment]
All the stack traces above are preceeded by this line:
```
Object St.Bin (0x55c14ad8fd00), has been already deallocated  - impossible to access to it. This might be caused by the fact that the object has been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs
```

**Comment by sustmi** [comment]
Hi @genodeftest . I am sorry I did not replied sooner.

I cannot really tell what is the cause of the issue. It looks like something inside Gnome Shell destroys objects (`createdActors`) that are supposed to be destroyed in `disable` function.

It looks like *gjs* (the JavaScript engine powering Gnome Shell) added some check recently that emmits a stack-trace if one tries to work with object that was already finalized. See https://gitlab.gnome.org/GNOME/gnome-shell/merge_requests/4 .

So maybe my extension always had this problem but it just did not show any symptoms. Or maybe Gnome Shell changed the way how the objects are freed.

They already changed the extensions API recently and now Gnome Shell calls `disable()` and `enable()` every time you lock/unlock screen.

The API is actually not documented really well. The only source that could be called a documentation is this: https://wiki.gnome.org/Projects/GnomeShell/Extensions/StepByStepTutorial
where they basically say that when Gnome Shell calls `disable()` your need to clean-up everything.

Looking into documentation for Clutter actors (https://developer.gnome.org/clutter/1.26/ClutterActor.html#clutter-actor-destroy), I can see that when parent actor is destroyed, all children actors are destroyed too. So I guess that may be the cause of the warnings.
The extensions adds the icons into WindowOverlay and when Gnome Shell destroys WindowOverlay, it destroys icons too so I should not destroy them again.

However these are just wild guesses. I will need to look deeper into the problem.

The problem is that right now I updated my *gjs* to v1.50.4-2 which does not log the stacktraces any more. See https://bugzilla.redhat.com/show_bug.cgi?id=1546059.
I tried to set `G_MESSAGES_DEBUG=all` as they recommend but it did not change anything.
So in order to reproduce the issue I will have to downgrade my *gjs* first. :/
I am not sure when I will find time for this.

**Comment by sustmi** [comment]
Hi @genodeftest . I tried to downgrade *gjs* to v1.50.1-1.fc27 but I cannot get the warnings any more.

If you still encounter any of this stactraces please send them here along with your version info (distro, gjs and gnome-shell).

**Comment by genodeftest** [comment]
After updating gjs to gjs-1.50.4-2.fc27.x86_64 (previous version: gjs-1.50.4-1.fc27.x86_64), I cannot reproduce this issue either because that version is [patched](https://src.fedoraproject.org/rpms/gjs/c/1663b7503563515abcf55a1ed0db45b52ac2108f?branch=f27) to avoid these stacktraces. I think these warnings were added way after 1.50.1-1.fc27. gjs-1.50.4-1.fc27 might be the only one with those warnings enabled.

Adding stuff to the environment is quite hard with gdm+gnome-shell because your `~/.bashrc` won't affect gnome-shell afaik. Running `less /proc/[gnome-shell user PID]/environ` should confirm that. There is another way of changing gnome-shell's environment, but I don't have it at hand right now.

**Comment by sustmi** [comment]
I tried:
```
export G_MESSAGES_DEBUG=all
gnome-shell --replace
```

(See https://gitlab.gnome.org/GNOME/gjs/merge_requests/71/diffs)

I get lot of debugging messages including:
```
(gnome-shell:569): Gjs-DEBUG: Object Meta.WindowActor (0x55cebe9ddf70), has been already finalized. Impossible to set any property to it.
== Stack trace for context 0x55cebc738170 ==
#0 0x55cebcbb3218 i   resource:///org/gnome/shell/ui/tweener.js:80 (0x7f2ca84ddef0 @ 82)
#1 0x55cebcbb3198 i   resource:///org/gnome/shell/ui/tweener.js:105 (0x7f2ca84df230 @ 36)
#2 0x55cebcbb3110 i   resource:///org/gnome/shell/ui/tweener.js:92 (0x7f2ca84df098 @ 52)
#3 0x7ffcddbc6840 b   resource:///org/gnome/gjs/modules/tweener/tweener.js:203 (0x7f2ca84e9cd0 @ 54)
#4 0x7ffcddbc6990 b   resource:///org/gnome/gjs/modules/tweener/tweener.js:332 (0x7f2ca84e9d58 @ 1626)
#5 0x7ffcddbc6a40 b   resource:///org/gnome/gjs/modules/tweener/tweener.js:345 (0x7f2ca84e9de0 @ 100)
#6 0x7ffcddbc6ad0 b   resource:///org/gnome/gjs/modules/tweener/tweener.js:360 (0x7f2ca84e9e68 @ 10)
#7 0x7ffcddbc6b50 I   resource:///org/gnome/gjs/modules/signals.js:126 (0x7f2ca84e2b38 @ 386)
#8 0x7ffcddbc6c00 b   resource:///org/gnome/shell/ui/tweener.js:208 (0x7f2ca84df808 @ 159)
#9 0x7ffcddbc6c60 I   resource:///org/gnome/gjs/modules/_legacy.js:82 (0x7f2ca84c2bc0 @ 71)
#10 0x7ffcddbc6d10 b   resource:///org/gnome/shell/ui/tweener.js:183 (0x7f2ca84df780 @ 20)
#11 0x7ffcddbc6de0 b   self-hosted:917 (0x7f2ca84ee5e8 @ 394)
```
but nothing from this extension.

UPDATE: Ok, I found gjs-1.50.4-1.fc27 RPM on some mirror and now I can reproduce the stacktraces so I will look into it.

**Comment by fritzmatias** [comment]
I am having the same issue, i can reproduce it under the context of using the browser (chrome & mozilla) after go in the calendar url (https://calendar.google.com/calendar/r) and accept the notifications.

**Comment by genodeftest** [comment]
Updated messages with GNOME 3.30 and gjs 1.54.1:

These stacktraces happen very often:

	#1   5589aa23d5f0 i   resource:///org/gnome/shell/ui/extensionSystem.js:70 (7f29acb549d0 @ 229)
	#2   5589aa23d570 i   resource:///org/gnome/shell/ui/extensionSystem.js:287 (7f29acb590d0 @ 13)
	#3   5589aa23d4a0 i   self-hosted:261 (7f29acfc1dc0 @ 223)
	#4   5589aa23d420 i   resource:///org/gnome/shell/ui/extensionSystem.js:284 (7f29acb54e50 @ 113)

The following two messages are printed by gnome-shell too, but less often and async to the traceback above:

	Object St.Bin (0x5589a8b22120), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.

and

	clutter_actor_destroy: assertion 'CLUTTER_IS_ACTOR (self)' failed

**Comment by sustmi** [comment]
The issues should be now fixed by dc1b6124d959c360e03af4b83f2d17f00921b3bf.
